base_portal_type/models/res_groups.py

```diff
@@ -24,51 +26,71 @@ def _update_user_groups_view(self):
         ][0]
         portal_group = self.env.ref("base.group_portal")
         internal_group = self.env.ref("base.group_user")
-        portal_groups = (
-            self.env["res.groups"]
-            .with_context(lang=None)
-            .search(
-                [
+
+        user_type_field_name = name_selection_groups(user_type_groups.ids)
+        non_public_attrs = str(
+            {
+                "readonly": [
                     (
-                        "category_id",
-                        "=",
-                        self.env.ref("base_portal_type.category_portal_type").id,
-                    ),
-                ]
-            )
-        )
-        user_type_field = name_selection_groups(user_type_groups.ids)
-        for node in arch.xpath("//group/field[@name='%s']" % user_type_field):
-            for group in portal_groups:
-                field_name = name_boolean_group(group.id)
-                for field_node in arch.xpath("//field[@name='%s']" % field_name):
-                    field_node.attrib["attrs"] = str(
-                        {
-                            "readonly": [
-                                (
-                                    user_type_field,
-                                    "not in",
-                                    (portal_group + internal_group).ids,
-                                )
-                            ],
-                        }
+                        user_type_field_name,
+                        "not in",
+                        [internal_group.id, portal_group.id],
                     )
-                node.addnext(
-                    etree.Element(
+                ],
+            }
+        )
+
+        portal_groups = self.env["res.groups"].search(
+            [
+                "|",
+                (
+                    "category_id",
+                    "=",
+                    self.env.ref("base_portal_type.category_portal_type").id,
+                ),
+                ("portal", "=", True),
+            ]
+        )
+
+        arch_root = arch.xpath("//field[@name='groups_id']")[0]
+        if len(arch_root) and portal_groups:
```

Comment:
```suggestion
        if not portal_groups:
            return result
        for arch_root in arch.xpath("//field[@name='groups_id']"):
```

no need to do the write below if we don't change anything, and `arch.xpath("//field[@name='groups_id']")[0]` raises if the xpath is empty for whatever reason

Reply:
true. I did it differently and moved the write() section into the if-block.
I think I now also understood that the whole function is called twice during install/upgrade - and the first time with the 'install_filename' in the context, which creates a placeholder view in the super() function, which therefore leaves us with the empty xpath result.
you didn't encounter that problem beforehand, because you where for-looping over the xpath results --> not doing a single loop in that case.

I didnt try, but I would even go as far as saying that our whole function _update_user_groups_view() can simply be skipped at the very top by checking if the 'install_filename' is in the context